pars-fmt/src/simple.rs

```diff
@@ -0,0 +1,294 @@
+                self.tokens.push(token);
+                self.pos = self.pos + end + 1;
+                self.state = State::Ready;
+                Ok(())
+            }
+            None => Err(FormatError::new(
+                self.source,
+                self.pos..self.source.len(),
+                "expected '}' found end of string",
+            )),
+        }
+    }
+
+    fn take_separator(&mut self, start_pos: usize) -> Result<(), FormatError> {
+        let end_pos = self.pos
+            + self.source.as_bytes()[self.pos..]
+                .iter()
+                .position(|b| b == &b'#')
+                .unwrap_or(self.source.len() - self.pos);
+
+        let sep_range = start_pos..end_pos;
+        assert!(start_pos < end_pos);
+        if let Some(Token::Separator(ref mut existing)) = self.tokens.last_mut() {
+            existing.end = end_pos;
+        } else {
+            self.tokens.push(Token::Separator(sep_range));
+        }
+        self.pos = end_pos;
+        self.state = State::Ready;
+        Ok(())
+    }
+
+    fn into_matcher(self) -> Result<FmtMatcher<'a>, FormatError> {
+        let tokens = self.tokens;
+        let lead_separator =
+            if let Some(Token::Separator(s)) = tokens.first() { Some(s.clone()) } else { None };
+
+        let mut fmt_fields = Vec::new();
+        let skip = if lead_separator.is_some() { 1 } else { 0 };
+        let mut iter = tokens.into_iter().skip(skip);
+        loop {
+            match (iter.next(), iter.next()) {
+                (Some(Token::Field(field)), Some(Token::Separator(sep))) => {
+                    fmt_fields.push((field, sep))
+                }
+                (Some(Token::Field(field)), None) => {
+                    fmt_fields.push((field, self.source.len()..self.source.len()))
+                }
+                (Some(Token::Field(rng)), _) | (Some(Token::Separator(rng)), _) => {
+                    return Err(FormatError::new(self.source, rng, "unexpected token"));
+                }
+                (None, None) => break,
+                (None, Some(_)) => unreachable!(),
+            }
+        }
+
+        Ok(FmtMatcher { source: self.source, lead_separator, fmt_fields, fields: Vec::new() })
+    }
+
+    #[cfg(test)]
+    fn resolve_tokens(&'a self) -> Vec<&'a str> {
+        self.tokens
+            .iter()
+            .map(|t| match t {
+                Token::Field(rng) => &self.source[rng.clone()],
+                Token::Separator(rng) => &self.source[rng.clone()],
+            })
+            .collect()
+    }
+}
+
+#[allow(dead_code)]
+pub struct FmtMatcher<'a> {
+    source: &'a str,
+    lead_separator: Option<Range<usize>>,
+    //NOTE: we represent all substrings as ranges of self.source, to avoid
+    //any unnecessary allocation.
+    fmt_fields: Vec<(Range<usize>, Range<usize>)>,
+    fields: Vec<String>,
+}
+
+#[allow(dead_code)]
+pub struct FmtMatch<'a, 'b> {
```

Comment:
I like this type. This is much better than returning a pair of lists.

Reply:
yea, and I think in upcoming work we can likely get rid of it completely, not totally sure yet though.